internal/server/data/identity.go

```diff
@@ -161,31 +163,61 @@ func DeleteIdentities(tx GormTxn, selectors ...SelectorFunc) error {
 		return err
 	}
 
-	ids := make([]uid.ID, 0)
-	for _, i := range toDelete {
-		ids = append(ids, i.ID)
+	ids, err := deleteReferencesToIdentities(tx, providerID, toDelete)
+	if err != nil {
+		return fmt.Errorf("remove identities: %w", err)
+	}
 
-		err := DeleteGrants(tx, DeleteGrantsOptions{BySubject: i.PolyID()})
-		if err != nil {
-			return err
+	if len(ids) > 0 {
+		return deleteAll[models.Identity](tx, ByIDs(ids))
+	}
+
+	return nil
+}
+
+func deleteReferencesToIdentities(tx GormTxn, providerID uid.ID, toDelete []models.Identity) (unreferencedIdentityIDs []uid.ID, err error) {
+	for _, i := range toDelete {
+		if err := DeleteAccessKeys(tx, DeleteAccessKeysOptions{ByIssuedForID: i.ID, ByProviderID: providerID}); err != nil {
+			return []uid.ID{}, fmt.Errorf("delete identity access keys: %w", err)
+		}
+		// if an identity does not have credentials in the Infra provider this won't be found, but we can proceed
+		credential, err := GetCredential(tx, ByIdentityID(i.ID))
```

Comment:
I'm not sure if this is correct. I believe we can get here from `data.DeleteProvider`, and if someone deletes an `okta` provider, won't this now delete the users infra credentials as well (because `data.DeleteIdentities` removes credentials by identity ID) ?

Reply:
Good catch, added a check to not delete credentials when not removing an identity from the infra provider and test for this case.